File: 2048.js
// 2048.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');

const COLORS = {
    reset: '\x1b[0m',
    bold: '\x1b[1m',
    red: '\x1b[91m',
    green: '\x1b[92m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    magenta: '\x1b[95m',
    cyan: '\x1b[96m',
    white: '\x1b[97m',
    bgRed: '\x1b[101m',
    bgGreen: '\x1b[102m',
    bgYellow: '\x1b[103m',
    bgBlue: '\x1b[104m',
    bgMagenta: '\x1b[105m',
    bgCyan: '\x1b[106m',
    bgWhite: '\x1b[107m'
};

function colorize(text, color) {
    return (COLORS[color] || '') + text + COLORS.reset;
}

const TILE_COLORS = {
    2: { bg: 'bgYellow', fg: 'black' },
    4: { bg: 'bgBlue', fg: 'white' },
    8: { bg: 'bgCyan', fg: 'black' },
    16: { bg: 'bgGreen', fg: 'white' },
    32: { bg: 'bgMagenta', fg: 'white' },
    64: { bg: 'bgRed', fg: 'white' },
    128: { bg: 'bgYellow', fg: 'black' },
    256: { bg: 'bgBlue', fg: 'white' },
    512: { bg: 'bgCyan', fg: 'black' },
    1024: { bg: 'bgMagenta', fg: 'white' },
    2048: { bg: 'bgRed', fg: 'white' },
    4096: { bg: 'bgGreen', fg: 'black' },
    8192: { bg: 'bgBlue', fg: 'white' },
};

function getTileColor(val) {
    return TILE_COLORS[val] || { bg: 'bgWhite', fg: 'black' };
}

const RECORD_FILE = path.join(os.homedir(), '.2048_records.json');

function loadRecords() {
    try {
        return JSON.parse(fs.readFileSync(RECORD_FILE, 'utf8'));
    } catch { return {}; }
}

function saveRecords(records) {
    fs.writeFileSync(RECORD_FILE, JSON.stringify(records, null, 2));
}

class Game2048 {
    constructor(size = 4) {
        this.size = size;
        this.board = Array.from({ length: size }, () => Array(size).fill(0));
        this.score = 0;
        this.history = [];
        this.records = loadRecords();
        this.best = this.records[String(size)] || 0;
    }

    addRandomTile() {
        const empty = [];
        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                if (this.board[i][j] === 0) empty.push([i, j]);
            }
        }
        if (empty.length === 0) return;
        const [i, j] = empty[Math.floor(Math.random() * empty.length)];
        this.board[i][j] = Math.random() < 0.9 ? 2 : 4;
    }

    compress(row) {
        const res = row.filter(x => x !== 0);
        while (res.length < this.size) res.push(0);
        return res;
    }

    merge(row) {
        const res = [];
        let add = 0;
        let i = 0;
        while (i < row.length) {
            if (i < row.length - 1 && row[i] === row[i + 1] && row[i] !== 0) {
                res.push(row[i] * 2);
                add += row[i] * 2;
                i += 2;
            } else {
                res.push(row[i]);
                i++;
            }
        }
        while (res.length < this.size) res.push(0);
        return [res, add];
    }

    move(dir) {
        const prevBoard = this.board.map(row => [...row]);
        const prevScore = this.score;
        let add = 0;

        const rotate = (board) => {
            const n = board.length;
            const result = Array.from({ length: n }, () => Array(n).fill(0));
            for (let i = 0; i < n; i++) {
                for (let j = 0; j < n; j++) {
                    result[j][n - 1 - i] = board[i][j];
                }
            }
            return result;
        };

        let board = this.board.map(row => [...row]);
        let times = 0;
        if (dir === 'w') times = 0;
        else if (dir === 'd') times = 1;
        else if (dir === 's') times = 2;
        else if (dir === 'a') times = 3;

        for (let t = 0; t < times; t++) board = rotate(board);

        for (let i = 0; i < this.size; i++) {
            const row = this.compress(board[i]);
            const [merged, scoreAdd] = this.merge(row);
            const finalRow = this.compress(merged);
            board[i] = finalRow;
            add += scoreAdd;
        }

        for (let t = 0; t < (4 - times) % 4; t++) board = rotate(board);

        if (board.some((row, i) => row.some((val, j) => val !== this.board[i][j]))) {
            this.history.push({ board: prevBoard.map(row => [...row]), score: prevScore });
            this.board = board;
            this.score += add;
            if (this.score > this.best) {
                this.best = this.score;
                this.records[String(this.size)] = this.best;
                saveRecords(this.records);
            }
            this.addRandomTile();
            return true;
        }
        return false;
    }

    undo() {
        if (this.history.length === 0) return false;
        const last = this.history.pop();
        this.board = last.board;
        this.score = last.score;
        return true;
    }

    isWin() {
        for (const row of this.board) {
            for (const val of row) {
                if (val >= 2048) return true;
            }
        }
        return false;
    }

    isGameOver() {
        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                if (this.board[i][j] === 0) return false;
                if (i < this.size - 1 && this.board[i][j] === this.board[i + 1][j]) return false;
                if (j < this.size - 1 && this.board[i][j] === this.board[i][j + 1]) return false;
            }
        }
        return true;
    }

    display() {
        console.clear();
        console.log(colorize(`🎮  2048  |  Размер ${this.size}×${this.size}  |  Счёт: ${this.score}  |  Лучший: ${this.best}`, 'bold'));
        console.log('─'.repeat(this.size * 8 + 1));
        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                const val = this.board[i][j];
                if (val === 0) {
                    process.stdout.write(colorize('      ', 'bgWhite') + ' ');
                } else {
                    const tc = getTileColor(val);
                    const text = String(val).padStart(6);
                    process.stdout.write(colorize(text, tc.bg) + ' ');
                }
                if (j < this.size - 1) process.stdout.write('│');
            }
            console.log();
            if (i < this.size - 1) console.log('─'.repeat(this.size * 8 + 1));
        }
        console.log('─'.repeat(this.size * 8 + 1));
        console.log('WASD — движение | U — отмена | Q — выход');
    }
}

async function main() {
    const args = process.argv.slice(2);
    let size = 4;
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '-s' && i + 1 < args.length) {
            size = parseInt(args[++i]);
        } else if (args[i] === '-h' || args[i] === '--help') {
            console.log('Usage: node 2048.js [options]\n  -s <N>   Size (4, 5, 6) default 4');
            return;
        }
    }
    if (size < 4 || size > 6) size = 4;

    const game = new Game2048(size);
    game.addRandomTile();
    game.addRandomTile();

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        terminal: true
    });

    readline.emitKeypressEvents(process.stdin);
    process.stdin.setRawMode(true);

    process.stdin.on('keypress', (str, key) => {
        if (key && key.name === 'q') {
            console.log(colorize('Игра сохранена. До встречи!', 'yellow'));
            process.exit(0);
        } else if (key && key.name === 'u') {
            if (game.undo()) console.log(colorize('Ход отменён.', 'yellow'));
            else console.log(colorize('Нечего отменять.', 'yellow'));
        } else if (str === 'w' || key && key.name === 'up') {
            game.move('w');
        } else if (str === 'a' || key && key.name === 'left') {
            game.move('a');
        } else if (str === 's' || key && key.name === 'down') {
            game.move('s');
        } else if (str === 'd' || key && key.name === 'right') {
            game.move('d');
        }
        game.display();
        if (game.isWin()) console.log(colorize('🎉 Поздравляем! Вы достигли 2048!', 'bold'));
        if (game.isGameOver()) console.log(colorize('💀 Игра окончена. Нажмите Q для выхода.', 'red'));
    });

    game.display();
}

main().catch(err => console.error(err));

File: 2048.rb
#!/usr/bin/env ruby
# 2048.rb
# encoding: UTF-8

require 'json'
require 'io/console'

COLORS = {
  reset: "\e[0m",
  bold: "\e[1m",
  red: "\e[91m",
  green: "\e[92m",
  yellow: "\e[93m",
  blue: "\e[94m",
  magenta: "\e[95m",
  cyan: "\e[96m",
  white: "\e[97m",
  bgRed: "\e[101m",
  bgGreen: "\e[102m",
  bgYellow: "\e[103m",
  bgBlue: "\e[104m",
  bgMagenta: "\e[105m",
  bgCyan: "\e[106m",
  bgWhite: "\e[107m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

TILE_COLORS = {
  2 => { bg: :bgYellow, fg: :black },
  4 => { bg: :bgBlue, fg: :white },
  8 => { bg: :bgCyan, fg: :black },
  16 => { bg: :bgGreen, fg: :white },
  32 => { bg: :bgMagenta, fg: :white },
  64 => { bg: :bgRed, fg: :white },
  128 => { bg: :bgYellow, fg: :black },
  256 => { bg: :bgBlue, fg: :white },
  512 => { bg: :bgCyan, fg: :black },
  1024 => { bg: :bgMagenta, fg: :white },
  2048 => { bg: :bgRed, fg: :white },
  4096 => { bg: :bgGreen, fg: :black },
  8192 => { bg: :bgBlue, fg: :white }
}

def get_tile_color(val)
  TILE_COLORS[val] || { bg: :bgWhite, fg: :black }
end

RECORD_FILE = File.join(Dir.home, '.2048_records.json')

def load_records
  return {} unless File.exist?(RECORD_FILE)
  JSON.parse(File.read(RECORD_FILE))
rescue
  {}
end

def save_records(records)
  File.write(RECORD_FILE, JSON.pretty_generate(records))
end

class Game2048
  attr_reader :size, :board, :score, :best, :history

  def initialize(size = 4)
    @size = size
    @board = Array.new(size) { Array.new(size, 0) }
    @score = 0
    @history = []
    @records = load_records
    @best = @records[size.to_s] || 0
    @rng = Random.new
  end

  def add_random_tile
    empty = []
    @size.times do |i|
      @size.times do |j|
        empty << [i, j] if @board[i][j] == 0
      end
    end
    return if empty.empty?
    i, j = empty.sample
    @board[i][j] = @rng.rand < 0.9 ? 2 : 4
  end

  def compress(row)
    res = row.reject { |x| x == 0 }
    res += [0] * (@size - res.length)
    res
  end

  def merge(row)
    res = []
    add = 0
    i = 0
    while i < row.length
      if i < row.length - 1 && row[i] == row[i + 1] && row[i] != 0
        res << row[i] * 2
        add += row[i] * 2
        i += 2
      else
        res << row[i]
        i += 1
      end
    end
    res += [0] * (@size - res.length)
    [res, add]
  end

  def rotate(board)
    board.transpose.map(&:reverse)
  end

  def move(dir)
    prev_board = @board.map(&:dup)
    prev_score = @score
    add = 0

    times = { 'w' => 0, 'd' => 1, 's' => 2, 'a' => 3 }[dir] || 0
    b = @board.map(&:dup)
    times.times { b = rotate(b) }

    b.size.times do |i|
      row = compress(b[i])
      merged, sc = merge(row)
      b[i] = compress(merged)
      add += sc
    end

    (4 - times) % 4.times { b = rotate(b) }

    if b != @board
      @history << { board: prev_board, score: prev_score }
      @board = b
      @score += add
      if @score > @best
        @best = @score
        @records[@size.to_s] = @best
        save_records(@records)
      end
      add_random_tile
      return true
    end
    false
  end

  def undo
    return false if @history.empty?
    last = @history.pop
    @board = last[:board]
    @score = last[:score]
    true
  end

  def is_win?
    @board.any? { |row| row.any? { |v| v >= 2048 } }
  end

  def is_game_over?
    @size.times do |i|
      @size.times do |j|
        return false if @board[i][j] == 0
        return false if i < @size - 1 && @board[i][j] == @board[i + 1][j]
        return false if j < @size - 1 && @board[i][j] == @board[i][j + 1]
      end
    end
    true
  end

  def display
    system('clear') || system('cls')
    puts colorize("🎮  2048  |  Размер #{@size}×#{@size}  |  Счёт: #{@score}  |  Лучший: #{@best}", :bold)
    puts '─' * (@size * 8 + 1)
    @size.times do |i|
      @size.times do |j|
        val = @board[i][j]
        if val == 0
          print colorize('      ', :bgWhite) + ' '
        else
          tc = get_tile_color(val)
          print colorize(val.to_s.rjust(6), tc[:bg]) + ' '
        end
        print '│' if j < @size - 1
      end
      puts
      puts '─' * (@size * 8 + 1) if i < @size - 1
    end
    puts '─' * (@size * 8 + 1)
    puts 'WASD — движение | U — отмена | Q — выход'
  end
end

def main
  size = 4
  i = 0
  while i < ARGV.length
    case ARGV[i]
    when '-s'
      size = ARGV[i+1].to_i
      i += 2
    when '-h', '--help'
      puts "Usage: ruby 2048.rb [options]\n  -s <N>   Size (4, 5, 6) default 4"
      return
    else
      i += 1
    end
  end
  size = 4 if size < 4 || size > 6

  game = Game2048.new(size)
  game.add_random_tile
  game.add_random_tile

  loop do
    game.display
    puts colorize("🎉 Поздравляем! Вы достигли 2048!", :bold) if game.is_win?
    puts colorize("💀 Игра окончена. Нажмите Q для выхода.", :red) if game.is_game_over?

    ch = STDIN.getch
    case ch
    when 'q', 'Q'
      puts colorize("Игра сохранена. До встречи!", :yellow)
      break
    when 'u', 'U'
      if game.undo
        puts colorize("Ход отменён.", :yellow)
      else
        puts colorize("Нечего отменять.", :yellow)
      end
    when 'w', 'W' then game.move('w')
    when 'a', 'A' then game.move('a')
    when 's', 'S' then game.move('s')
    when 'd', 'D' then game.move('d')
    end
  end
end

main if __FILE__ == $0
